onyo/lib/pseudokeys.py

```diff
@@ -0,0 +1,101 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from functools import partial
+from typing import TYPE_CHECKING
+
+
+if TYPE_CHECKING:
+    from typing import (
+        Callable,
+        Dict,
+    )
+    from onyo.lib.items import Item
+
+
+@dataclass()
+class PseudoKey:
+    """"Defines a pseudo-key implementation."""
+    description: str
+    """description of what that pseudo-key is delivering"""
+    implementation: Callable
+    """Called by an `Item` when a pseudo-keys value is still undefined.
+
+    This Callable is expected to have a single parameter of type `Item`.
+    """
+
+
+def delegate(self: Item, attribute: str, *args, **kwargs):
+    # This is to avoid circular imports ATM.
+    return self.__getattribute__(attribute)(*args, **kwargs)
+
+
+PSEUDO_KEYS: Dict[str, PseudoKey] = {
+    'onyo.path.absolute': PseudoKey(description="Absolute path of the item.",
+                                    implementation=partial(delegate, attribute='get_path_absolute')
+                                    ),
+    'onyo.path.relative': PseudoKey(description="Path of the item relative to the repository root.",
+                                    implementation=partial(delegate, attribute='get_path_relative')
+                                    ),
+    'onyo.path.parent': PseudoKey(description="Path of the directory the item is in, relative to the repository root.",
+                                  implementation=partial(delegate, attribute='get_path_parent')
+                                  ),
+    'onyo.path.file': PseudoKey(description="Path to the file containing an asset's YAML."
+                                            "Different from 'onyo.path.relative' in case of an asset directory.",
+                                implementation=partial(delegate, attribute='get_path_file')
+                                ),
+    'onyo.is.asset': PseudoKey(description="Is the item an asset.",
+                               implementation=partial(delegate, attribute='is_asset')
+                               ),
+    'onyo.is.directory': PseudoKey(description="Is the item a directory.",
+                                   implementation=partial(delegate, attribute='is_directory')
+                                   ),
+    'onyo.is.template': PseudoKey(description="Is the item a template.",
+                                  implementation=partial(delegate, attribute='is_template')
+                                  ),
+    'onyo.is.empty': PseudoKey(description="Is the directory empty. <unset> if the item is not a directory.",
+                               implementation=partial(delegate, attribute='is_empty')
+                               ),
+}
+r"""Pseudo-Keys are key names that are addressable but not in asset content.
```

Comment:
I would adjust this just a little bit:

> Pseudo-Keys are key names that are addressable but not in asset content.

to 

> Pseudo-Keys are keys that are addressable but not written to disk as YAML.

Or something like that. it actually is a bit difficult to phrase right, since `onyo.is.directory` can be written to disk (just not as YAML). And using the word "file" is correct, but gets confusing WRT asset directories.

My goal was to make it clear that it was on-disk. The asset content in-memory clearly has these keys.

Reply:
True. It's actually just taken from the previous `PSEUDO_KEYS` definition in `consts.py`.

> since onyo.is.directory can be written to disk (just not as YAML).

Yeah, and I think we said, we'd allow for several pseudo-keys to be set in templates. So, it's "on disc" and "in YAML", just not in an asset file.
So,

"Pseudo-Keys are key names that are addressable but not written to asset files."?